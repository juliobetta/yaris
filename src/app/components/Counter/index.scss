$background: #ededeb;

.Counter {
  background-color: $background;
  padding: 1rem;
  display: flex;
  justify-content: space-between;
  align-items: center;

  > span {
    font-size: 4rem;
  }
}
