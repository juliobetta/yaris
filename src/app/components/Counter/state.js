export function initialState(initialValue = 1) {
  return { counter: initialValue };
}

export function increment(state, step = 1) {
  return { counter: (state.counter + step) };
}

export function decrement(state, step = 1) {
  return { counter: (state.counter - step) };
}

export function reset(initialValue = 1) {
  return { counter: initialValue };
}
