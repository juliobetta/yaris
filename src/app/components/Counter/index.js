import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { initialState, increment, decrement, reset } from './state';
import Styles from './index.less';


class Counter extends PureComponent {
  static propTypes = {
    initialValue: PropTypes.number,
    resetLabel: PropTypes.string,
    step: PropTypes.number
  }

  static defaultProps = {
    initialValue: 1,
    step: 1
  }

  state = initialState(this.props.initialValue);

  render() {
    const { initialValue, step, resetLabel } = this.props;

    return (
      <div className={Styles.Counter}>
        <div>
          <button onClick={this.handleDecrement(initialValue, step)}>â€“</button> { ' ' }
          <button onClick={this.handleIncrement(step)}>+</button> { ' ' }
          <button className="reset" onClick={this.handleReset(initialValue)}>{resetLabel}</button>
        </div>

        <span className="counter">{this.state.counter}</span>
      </div>
    );
  }

  handleIncrement = step => () => this.setState(state => increment(state, step));
  handleDecrement = (initialValue, step) => () => {
    const newState = decrement(this.state, step);

    if(initialValue <= newState.counter) {
      this.setState({ ...this.state, ...newState });
    }
  };
  handleReset = initialValue => () => this.setState(() => reset(initialValue))
}


export default Counter;
