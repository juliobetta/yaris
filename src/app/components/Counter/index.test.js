import React from 'react';
import Counter from './index';

describe('<Counter>', () => {
  it('renders correctly', () => {
    const wrapper = mount(<Counter resetLabel="reset" />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
