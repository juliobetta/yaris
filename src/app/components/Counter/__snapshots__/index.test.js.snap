// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter> renders correctly 1`] = `
<Counter
  initialValue={1}
  resetLabel="reset"
  step={1}
>
  <div
    className="Counter"
  >
    <div>
      <button
        onClick={[Function]}
      >
        â€“
      </button>
       
       
      <button
        onClick={[Function]}
      >
        +
      </button>
       
       
      <button
        className="reset"
        onClick={[Function]}
      >
        reset
      </button>
    </div>
    <span
      className="counter"
    >
      1
    </span>
  </div>
</Counter>
`;
